plugins/Amplifier/Amplifier.cpp

```diff
@@ -65,17 +67,17 @@ bool AmplifierEffect::processAudioBuffer(sampleFrame* buf, const fpp_t frames)
 	const float d = dryLevel();
 	const float w = wetLevel();
 
-	const ValueBuffer* volumeBuf = m_ampControls.m_volumeModel.valueBuffer();
-	const ValueBuffer* panBuf = m_ampControls.m_panModel.valueBuffer();
-	const ValueBuffer* leftBuf = m_ampControls.m_leftModel.valueBuffer();
-	const ValueBuffer* rightBuf = m_ampControls.m_rightModel.valueBuffer();
+	const auto* volumeBuf = m_ampControls.m_volumeModel.valueBuffer();
+	const auto* panBuf = m_ampControls.m_panModel.valueBuffer();
+	const auto* leftBuf = m_ampControls.m_leftModel.valueBuffer();
+	const auto* rightBuf = m_ampControls.m_rightModel.valueBuffer();
 
-	for (fpp_t f = 0; f < frames; ++f)
+	for (size_t f = 0; f < frames; ++f)
 	{
-		const float volume = (volumeBuf ? volumeBuf->value(f) : m_ampControls.m_volumeModel.value()) * 0.01f;
-		const float pan = (panBuf ? panBuf->value(f) : m_ampControls.m_panModel.value()) * 0.01f;
-		const float left = (leftBuf ? leftBuf->value(f) : m_ampControls.m_leftModel.value()) * 0.01f;
-		const float right = (rightBuf ? rightBuf->value(f) : m_ampControls.m_rightModel.value()) * 0.01f;
+		const float volume = (volumeBuf ? volumeBuf->at(std::clamp<std::size_t>(f, 0, volumeBuf->size())) : m_ampControls.m_volumeModel.value()) * 0.01f;
+		const float pan = (panBuf ? panBuf->at(std::clamp<std::size_t>(f, 0, panBuf->size())) : m_ampControls.m_panModel.value()) * 0.01f;
+		const float left = (leftBuf ? leftBuf->at(std::clamp<std::size_t>(f, 0, leftBuf->size())) : m_ampControls.m_leftModel.value()) * 0.01f;
+		const float right = (rightBuf ? rightBuf->at(std::clamp<std::size_t>(f, 0, rightBuf->size())) : m_ampControls.m_rightModel.value()) * 0.01f;
```

Comment:
The `at` method is already bounds checked, so if we're confident `f` should never be out-of-bounds, we could either use `at` without the `std::clamp`, or for even greater efficiency, the unchecked `[]` operator (possibly with some `assert`s beforehand so that debug builds are bounds-checked).

Reply:
I believe it's better to care about safety and use `at` here. I don't really know much about asserts and don't want to bring in behaviour changes or out of bounds access issues.

Reply:
If there's an out of bound access, its a bug, and we would get a crash. Using `at` would also crash in this instance anyways, but it throws an `out_of_range` exception instead of straight crashing.

Besides, I don't think my suggestion uses `at`? It uses `valueAt`, which is not like the `at` function for `std::vector`.

Reply:
It would be more efficient and just as safe to use `std::clamp` with `[]` compared to what you're doing currently.

Using `at` with `std::clamp` would do bounds checking twice.

Using `at` _without_ `std::clamp` would throw an exception if it ever tried to access an out-of-bounds element. I'm not sure if `f` is supposed to always be within bounds or not. If it is, using `std::clamp` could be covering up a bug that would be detected via a thrown exception if we used plain `at` instead. If it isn't, it should stay as it is, but with `[]` instead of `at` for efficiency.

Reply:
`assert` is a macro that checks if an expression is true only for Debug builds. The expression is not compiled at all during Release builds.

So this code would be bounds checked in Debug builds (for better safety / documenting assumptions) but not in Release builds (for better efficiency):
```c++
assert(0 < f && f < buf.size()); // in Debug builds, calls std::abort() if condition is false
float value = buf[f];
```

Reply:
> Besides, I don't think my suggestion uses at? It uses valueAt, which is not like the at function for std::vector.

Sorry, I wasn't commenting on your suggestion - just the same section of code from the PR

Reply:
I used valueAt like saker said, there's no bounds checking here, if anything, this might as well crash.